feat: support deprecated field for OpenAPI parameters

lachlancooper opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to gRPC-Gateway here: https://github.com/grpc-ecosystem/grpc-gateway/blob/main/CONTRIBUTING.md

Happy contributing!

-->

#### References to other Issues or PRs

<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #2906

#### Have you read the [Contributing Guidelines](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/CONTRIBUTING.md)?

#### Brief description of what is fixed or changed

The `protoc-gen-openapiv2` generator now inspects each proto field’s deprecated option when building query/path parameters, and if set, writes the OpenAPI `deprecated` flag on the corresponding parameter object. This allows the proto-level deprecation signal to flow straight into the emitted openapiv2 document.

#### Other comments

johanbrandhorst reviewed (changes requested):
Was this all it took 😂? I ran the CI jobs, lets see if there are any failures. In any case, could you try applying a "deprecated" option to one of our protobuf fields to check that it works? Also, could we try to add an annotation option in addition to the protobuf option? It would be nice if users could deprecate OpenAPI fields independently of the protobuf deprecation. See CONTRIBUTING.md for steps on how to regenerate our test files.

lachlancooper commented:
I've implemented a `deprecated` annotation option under `field_configuration`. I've also updated the docs to cover the two different ways to mark a parameter as deprecated.

Since this whole PR is effectively a breaking change, should this all be behind a new plugin option, similar to `enable_rpc_deprecation`?

johanbrandhorst commented:
I agree that it is a breaking change, but there's also a case for this being a "bug fix". If someone had their protobuf fields marked deprecated and it didn't show up in their OpenAPI specs, that could be intentional, but it could also be an unfortunate shortcoming. As long as we make it easy for users to discover how to revert back to the old behavior (e.g., set the openapi field configuration deprecation option to false), it might be OK? What do you think? Are we more likely to delight users or annoy users with this?

lachlancooper commented:
> As long as we make it easy for users to discover how to revert back to the old behavior (e.g., set the openapi field configuration deprecation option to false)

As it stands right now, this won't be possible. The new field config option is a simple `bool` without presence, so if it's not explicitly set to `true`, then it defaults to `false` and we can't distinguish that from an explicit `false`.

In other words, the logic between the annotation and the proto option is currently OR, but this means a (arguably) breaking change with no way for users to override it. But if we make the logic AND, then users would need to add the annotation to every single field they want to deprecate, which is obviously not sensible.

I guess this is an argument in favour of adding a new top-level plugin option?
